Images can be saved, undo & redo is working

Onetchou opened the pull request:
Fixes #1100 
Fixes #1098

Onetchou committed:
feat: DraftImage struct updated

Onetchou committed:
feat: xml attributes updated for the image feature

Onetchou committed:
feat: images can be parsed from the sm2d file

Onetchou committed:
fix: lib include order fixed

Onetchou committed:
fix: loop between image and handles removed

Onetchou committed:
feat: images can now be saved, undo & redo are also working

Onetchou committed:
feat: info popup removed when image is imported

Onetchou committed:
refactor: code simplification in image_tool

Onetchou committed:
fix: image z order works again

Onetchou committed:
fix: ts files updated

Onetchou commented:
Since this PR is quite big and adds a lot of things it would be great if someone else than me could also test it to check if there's no unexpected crashes of the software in random situations 😉.

DSCaskey reviewed (commented):
, m_isHovered(false)
    , m_coloredHandles(false)
    
    I get the desire to change the flag variable name, but it makes no sense. For one it doesn't match the condition that you're testing for... that is are you in a hover event? Secondly the handles are always colored... they're either darkgrey or red. Thirdly isHovered is used in all the other tools, so now we're breaking continutity by renaming this flag. 
    
    I think what I would do, since that's the condition we're really checking, is name this flag something like m_handleIsHighlighted so that
    
    if (m_handleIsHighlighted){} ... reads better than if (isHovered){}
    
    Then to be consistent I would then also address the other isHovered's at some point like:
    
    m_imageIsHighlighted, m_lineIsHighlighted, m_scenePointIsHighlighted, etc...

DSCaskey commented:
There's several issues I found if loading a pattern or redo'ng and an image file is deleted while working - not likely to happen, but possible. First is the dialog that pops up tells the user the file can not be read and it may be corrupt. This is not the case, but rather the file is missing Big difference. 

![Screenshot 2024-07-31 222704](https://github.com/user-attachments/assets/887e7a69-09ac-4f22-b641-d5c10edcdbaa)

The big issue is that the after you click OK, the app crashes. When reading, or redo'ing, need to check if the file exists before continuing... if NOT we need a new Error message box stating the file is not found... along the lines like not finding a measurement file when loading a pattern. A bonus would be the option to "find" a / the file. Again, along the lines of finding a missing measurement file.

DSCaskey commented:
> const QString VPatternConverter::PatternMaxVerStr = QStringLiteral("0.7.0");
> const QString VPatternConverter::CurrentSchema    = QStringLiteral("://schema/pattern/v0.7.0.xsd");
> const QString VPatternConverter::PatternMaxVerStr = QStringLiteral("0.8.0");
> const QString VPatternConverter::CurrentSchema    = QStringLiteral("://schema/pattern/v0.8.0.xsd");

Is there a reason we went from 0.7.0 to 0.8.0 and skipped all the schemas in between? 

For consistency, the next locical bump in schema should be v0.7.1.xsd.

Onetchou commented:
> I get the desire to change the flag variable name, but it makes no sense. For one it doesn't match the condition that you're testing for... that is are you in a hover event? [...] Thirdly isHovered is used in all the other tools, so now we're breaking continutity by renaming this flag. 

I changed the name of the flag because I added a condition that is not related to a hover event:

![image](https://github.com/user-attachments/assets/7f011bde-220a-4fc0-8ec8-eb17d61757ed)

You can be over the handle but if you unclick it will unset the flag, even if it's still hovered...

> Secondly the handles are always colored... they're either darkgrey or red.
> I think what I would do, since that's the condition we're really checking, is name this flag something like m_handleIsHighlighted

😆 ok I'll change it if you want, no problem. But I don't think this is really a problem since it's quite understandable I think. Because if we want to use proper scientific definitions everywhere, one could argue that black is not a color and we'll have a lot of things to change in the software to correct this 😉 

> Then to be consistent I would then also address the other isHovered's at some point like:
> 
> m_imageIsHighlighted, m_lineIsHighlighted, m_scenePointIsHighlighted, etc...
> ```

I think that isHovered is perfectly fine everywhere else. I changed it here only because I changed the behavior of the flag: it doesn't reflect anymore the fact that the handle is hovered or not.

Onetchou commented:
> There's several issues I found if loading a pattern or redo'ng and an image file is deleted while working - not likely to happen, but possible. First is the dialog that pops up tells the user the file can not be read and it may be corrupt. This is not the case, but rather the file is missing Big difference.
> 
> ![Screenshot 2024-07-31 222704](https://private-user-images.githubusercontent.com/31944718/355556986-887e7a69-09ac-4f22-b641-d5c10edcdbaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyMTk4MDYsIm5iZiI6MTcyMzIxOTUwNiwicGF0aCI6Ii8zMTk0NDcxOC8zNTU1NTY5ODYtODg3ZTdhNjktMDlhYy00ZjIyLWI2NDEtZDVjMTBlZGNkYmFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA5VDE2MDUwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhZjZjMGY1ZDUxYTk0MTc5MzlkYTQ2MmNmYmRmYTViMzBlZjQzMTc4N2FhYTRiMTkyMjczMGEwOGY4YzJiOWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sXeHB0imu7NDMuXMfdw9HHJ4Rc6aNAqZwAE2v6QgV38)
> 
> The big issue is that the after you click OK, the app crashes. When reading, or redo'ing, need to check if the file exists before continuing... if NOT we need a new Error message box stating the file is not found... along the lines like not finding a measurement file when loading a pattern.

I was so focused to make the undo and redo work that I completely forgot to address the cases where the file is missing: I'll take a look at it, thanks for noticing it!



> A bonus would be the option to "find" a / the file. Again, along the lines of finding a missing measurement file.

Yes, great idea, I'll implement it.

Onetchou commented:
> Is there a reason we went from 0.7.0 to 0.8.0 and skipped all the schemas in between?

Because I thought it was a major change since new items are added, this is not only new attributes of existing items:

![image](https://github.com/user-attachments/assets/298db50d-ccd8-46c1-b095-66f55847965c)

DSCaskey commented:
>You can be over the handle but if you unclick it will unset the flag, even if it's still hovered...

But the handles are always colored like I said. m_coloredHandles = false implies the handles are no longer colored. BTW... using "m_coloredHandles" implies you're coloring the handles NOT just this handle. 

>I think that isHovered is perfectly fine everywhere else. I changed it here only because I changed the behavior of the flag: it doesn't reflect anymore the fact that the handle is hovered or not.

Again, the condition that you're checking is whether to highlight the handle Red or not... which was my point of m_handleIsHighlighted.  If it makes more sense... you could use m_highlightedHandle. Either one in English indicates the handle is highlighted... semantically different than coloredHandles or handlesAreColored. Yes, a highlighted handle is colored, but not every colored handle is highlighted. If you wanted to use m_redColoredHandle... then you'd be more correct, but still miss the WHY it's red. 

 painter->setBrush(m_highlightedHandle ? QColor(Qt::red) : QColor(Qt::darkGray));

> Because I thought it was a major change since new items are added, this is not only new attributes of existing items:

Fair enough.  I don't know why that's even there. Seems like it applies more to a build version, not a schema. The schema has nothing directly to do with a release. We could build a 1000 releases and patches without ever changing the schema. I suspect that RT just cut and pasted code from somewhere else not understanding the comments are not applicable in the case of the schemas. In fact the schema could have just as well been coded as a single digit, and it would have been easier. Like v001.xsd, v002.xsd... v100.xsd. If there is harder way to do things, that's what RT does. 

That being said if it's not that big a hassle I'd like to see this schema as v0.7.1.xsd just so down the line someone else does wonder what happened to v0.7.1.xsd to v0.7.9.xsd... or worse try to use one of them.

Onetchou committed:
refactor: m_coloredHandles is now m_handleIsHighlighted

Onetchou committed:
refactor: schema 0.8.0 is now 0.7.1

Onetchou committed:
feat: handling of broken image paths

Onetchou commented:
Another error to solve (memo for myself):
New pattern > Add image > Undo > Redo > Add a point (endline tool) >
![image](https://github.com/user-attachments/assets/c13850ad-30a8-4b29-950e-417f546f4413)

Onetchou commented:
> When reading, or redo'ing, need to check if the file exists before continuing... if NOT we need a new Error message box stating the file is not found... along the lines like not finding a measurement file when loading a pattern. A bonus would be the option to "find" a / the file. Again, along the lines of finding a missing measurement file.

Done 👌 

![image](https://github.com/user-attachments/assets/e50294cd-bdd7-4930-aa0e-bf1a6f5e0793)



> I think what I would do, since that's the condition we're really checking, is name this flag something like m_handleIsHighlighted

Done 👌 



> That being said if it's not that big a hassle I'd like to see this schema as v0.7.1.xsd just so down the line someone else does wonder what happened to v0.7.1.xsd to v0.7.9.xsd... or worse try to use one of them.

Done 👌 

😊

DSCaskey reviewed (approved):
Looks good to me.

DSCaskey committed:
Merge branch 'develop' into save-images

Signed-off-by: DSCaskey <31944718+DSCaskey@users.noreply.github.com>

Onetchou commented:
@DSCaskey I've set this PR as a "draft PR", because it was not ready to be merged, there's still this important issue to be solved :


> Another error to solve (memo for myself):
> New pattern > Add image > Undo > Redo > Add a point (endline tool) >
> ![image](https://github.com/user-attachments/assets/c13850ad-30a8-4b29-950e-417f546f4413)
> 
>
